### ISSUE TITLE
Update to embedded_hal 1.0.0

### ISSUE THREAD

**Issue by rappet** [smart-leds-rs/ws2812-timer-delay-rs#5]
I want to update this crate to `embedded_hal 1.0.0`.

The problem here is, that `embedded_hal 1.0.0` does not specify a timer anymore.
Is there any good alternative to use timers in an abstract way?

If not, should I create a pull request using `delay_ns` (tested, it works in [my unpolished fork](https://github.com/rappet/ws2812-delay-rs), but it is probably not in the scope of this crate), or should I just keep it as a fork?

**Comment by ianrrees** [comment]
Thanks for this!  `DelayNs` looks appropriate to me - anything I can do to help turn this in to a PR?

**Comment by rappet** [comment]
You can do some hardware testing to ensure it works in different environments. The timings I choose are a result of guessing and experimenting on my side, so they can be improved.

As far as I remember only slow mode did really work for me reliable on a 48 MHz CH32V003.

**Comment by rappet** [comment]
Also feel free to create a PR with my code if I don't find the time for it.

**Comment by ianrrees** [comment]
Cheers!  I've put this on my to-do list, but have a couple other items I'd like to wrap up first.

One thing I'm thinking is to provide a parameter (either in a constructor, or as a const generic) for the amount of time that a GPIO set and overhead for a timer wait takes, because that can change considerably depending on the microcontroller/HAL/profile.  This way, we could keep this crate generic, but tune it at instantiation.  Eg, if we want a 350us pulse, we might do something in the crate like:
``` Rust
self.gpio.set_high();
self.timer.delay_us(350 - self.overhead_us);
```

**Comment by david-sawatzke** [comment]
As you already mentioned, `embedded-hal` v1.0 doesn't really have a compatible API anymore. Previously we had implementations also based on bitbanging with nops (https://github.com/smart-leds-rs/ws2812-nop-samd21), but this was very fiddly and not really maintainable due to changes on different compiler versions. I think the `delay_*` methods will probably have similar issue, the best option is probably forking this crate and modifying it for a specific hal/MCU.

In my opinion e-h v1.0 doesn't provide enough functionality for this crate to function and is thus not supported.

**Comment by ianrrees** [comment]
Yeah, that's fair enough @david-sawatzke !  `DelayNs` provides delays of >= some time, so it doesn't really provide enough of a timing guarantee to reliably drive a WS2812.

I'm mainly interested in the SAMD21 and SAMD51 micros, and think that on those it's probably best to either use the SPI or TCC peripherals to drive WS2182s.  The embedded-hal SPI trait - quite reasonably, I think - doesn't seem to provide the level of control that would be required to drive a WS2182, so in practice I think this means that a WS2182 driver needs to be more-or-less specific to the target microcontroller.
